### PR TITLE
KAFKA-4785: Records from internal repartitioning topics should always use RecordMetadataTimestampExtractor

### PR DESCRIPTION
**Pull request by jeyhunkarimov**

### CODE REVIEW

**Comment by mjsax** [comment]
Thanks for the PR. The JIRA also applies to all internally created topics. Cf. `KStreamImpl#createReparitionedSource()` -- there might be other places in the code, too.

**Comment by jeyhunkarimov** [comment]
Thanks for comment. After a bit search, I think the method you mentioned is the most general one (and the only one I think) for internal topic repetitions

**Review by mjsax — APPROVED** [review]
LGTM

Call for second review @dguy @bbejeck @enothereska

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java [hunk 1]
```diff
@@ -149,6 +153,23 @@ public Integer apply(Integer value1, Integer value2) {
     }
 
     @Test
+    public void testTimestampExtractorsWithThrough() {
+        final KStreamBuilder builder = new KStreamBuilder();
+        KStream<String, String> stream1 = builder.stream(stringSerde, stringSerde, "topic-1", "topic-2");
+        KStream<String, String> stream2 = builder.stream(stringSerde, stringSerde, "topic-3", "topic-4");
+
+        stream1.to("topic-5");
+        stream2.through("topic-6");
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java:166]: Maybe we can extend this test, such that an internal repartitioning topic in created?

**Review by enothereska — APPROVED** [review]

**Review by bbejeck — APPROVED** [review]

**Review by dguy — COMMENTED** [review]
As @mjsax suggested - it would be good to have a test that proves this works for repartition topics.

**Comment by jeyhunkarimov** [comment]
@mjsax Does the recent update solve the limitation about tests?

**Review by mjsax — APPROVED** [review]
LGTM.

Call for final review and merging @guozhangwang

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5594/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5608/
Test PASSed (JDK 7 and Scala 2.11).

**Review by dguy — APPROVED** [review]
LGTM

**Pull request closed without merging**